### PR TITLE
[Port rc.2.0] fix(id-compressor): Prevent resubmission of ID allocation ops

### PR DESCRIPTION
**Pull request by kian-thompson**
### DO NOT MERGE UNTIL CONFIRMED

https://github.com/microsoft/FluidFramework/pull/21043

### Reviewer Guidance

There were a ton of commits since the commit we're interested in. I tried to take the minimum subset of changes and hot fix certain merge conflicts around `Outbox` and `BatchManager`

### CODE REVIEW

**Comment by vladsud** [comment]
@taylorsw04, do we need to port it to RC2? What is the scenario where it's needed?

I have a feeling that the only use (before 2.0) is for Loop & others to leverage it for short DDS/Datastore IDs, and that requires https://github.com/microsoft/FluidFramework/pull/20531, https://github.com/microsoft/FluidFramework/pull/19859.
The former was merged on April 8th, which looks like after RC2 was snapped.
Maybe we do not care (as enabling it in new files is good starting point).
Just trying to understand what we are trying to achieve...

**Comment by taylorsw04** [comment]
@vladsud this fix prevents a document corruption bug; any clients without this fix that use the compressor (for anything, so ST or datastore IDs) and hit the resubmit path with allocation ops in pending state will make the document permanently fail on load.

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts [hunk 1]
```diff
@@ -113,10 +113,10 @@ export class Outbox {
...
 		const softLimit = isCompressionEnabled ? Infinity : this.defaultAttachFlowSoftLimitInBytes;
 
-		this.attachFlowBatch = new BatchManager({ hardLimit, softLimit });
-		this.mainBatch = new BatchManager({ hardLimit });
-		this.blobAttachBatch = new BatchManager({ hardLimit });
-		this.idAllocationBatch = new BatchManager({ hardLimit });
+		this.attachFlowBatch = new BatchManager({ hardLimit, softLimit, canRebase: true });
+		this.mainBatch = new BatchManager({ hardLimit, canRebase: true });
+		this.blobAttachBatch = new BatchManager({ hardLimit, canRebase: true });
+		this.idAllocationBatch = new BatchManager({ hardLimit, canRebase: true });
```
Comment (taylorsw04) [inline packages/runtime/container-runtime/src/opLifecycle/outbox.ts:119]:
```suggestion
		this.idAllocationBatch = new BatchManager({ hardLimit, canRebase: false });
```

Comment (taylorsw04) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts:119]: this PR will not fix the bug unless this boolean is set :)

**Review by taylorsw04 — COMMENTED** [review]
If my one comment is resolved, this should be good to merge.

**Review by vladsud — APPROVED** [review]
Runtime changes look good. That said, unless we have direct feedback from customers or data suggesting this is required, I'd suggest not to port, to reduce risk.
I.e. I do not think we have a need from internal customers (we have not enabled shorter IDs in runtime yet and can wait extra 1-2 months). So it all depends on - if we have 3rd parties complaining about these issues for ST.

**Pull request closed without merging**